foreign-key: add note about inline references (#13056)
pingcap/docs: foreign-key.md

### Compatibility with MySQL

When you create a foreign key without specifying a name, the name generated by TiDB is different from that generated by MySQL. For example, the foreign key name generated by TiDB is `fk_1`, `fk_2`, and `fk_3`, while the foreign key name generated by MySQL is `table_name_ibfk_1`, `table_name_ibfk_2`, and `table_name_ibfk_3`.

Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of `FOREIGN KEY` definitions are checked and enforced. For example:

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT REFERENCES parent(id)
);

show create table child;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| child | CREATE TABLE `child` (                                      |
|       |   `id` int(11) DEFAULT NULL,                                |
|       |   `pid` int(11) DEFAULT NULL                                |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |

> +-------+-------------------------------------------------------------+

Author: ````suggestion
+-------+-------------------------------------------------------------+
```
````

### Compatibility with MySQL

When you create a foreign key without specifying a name, the name generated by TiDB is different from that generated by MySQL. For example, the foreign key name generated by TiDB is `fk_1`, `fk_2`, and `fk_3`, while the foreign key name generated by MySQL is `table_name_ibfk_1`, `table_name_ibfk_2`, and `table_name_ibfk_3`.

Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of `FOREIGN KEY` definitions are checked and enforced. For example:

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT REFERENCES parent(id)
);

> show create table child;

Author: ```suggestion
SHOW CREATE TABLE child;
```

+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| child | CREATE TABLE `child` (                                      |
|       |   `id` int(11) DEFAULT NULL,                                |
|       |   `pid` int(11) DEFAULT NULL                                |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+

> ```

Reviewer: In in uppercase, `SHOW CREATE TABLE` statement output indicates that no foreign keys exist in table child.

Please add this into doc, I miss it before, sorry...

Reviewer: The last three line should be this:

<img width="792" alt="image" src="https://user-images.githubusercontent.com/26020263/228557540-b97772ea-4590-4ac8-acac-723946e1a054.png">

Reviewer: you just delete the ``` which we should keep it.

> The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.

Author: ````suggestion
```
The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.
````

Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of `FOREIGN KEY` definitions are checked and enforced. For example:

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT REFERENCES parent(id)
);

SHOW CREATE TABLE child;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| child | CREATE TABLE `child` (                                      |
|       |   `id` int(11) DEFAULT NULL,                                |
|       |   `pid` int(11) DEFAULT NULL                                |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

> The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.

Reviewer: ```suggestion

The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.
```

### Compatibility with MySQL

When you create a foreign key without specifying a name, the name generated by TiDB is different from that generated by MySQL. For example, the foreign key name generated by TiDB is `fk_1`, `fk_2`, and `fk_3`, while the foreign key name generated by MySQL is `table_name_ibfk_1`, `table_name_ibfk_2`, and `table_name_ibfk_3`.

> Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of `FOREIGN KEY` definitions are checked and enforced. For example:
> 
> ```sql
> CREATE TABLE parent (
>     id INT KEY
> );
> 
> CREATE TABLE child (
>     id INT,
>     pid INT REFERENCES parent(id)
> );
> 
> SHOW CREATE TABLE child;
> +-------+-------------------------------------------------------------+
> | Table | Create Table                                                |
> +-------+-------------------------------------------------------------+
> | child | CREATE TABLE `child` (                                      |
> |       |   `id` int(11) DEFAULT NULL,                                |
> |       |   `pid` int(11) DEFAULT NULL                                |
> |       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
> +-------+-------------------------------------------------------------+
> ```
> 
> The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.

Reviewer 2: ````suggestion
Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of `FOREIGN KEY` definitions are checked and enforced. The following example creates a foreign key constraint using a `REFERENCES` clause:

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT REFERENCES parent(id)
);

SHOW CREATE TABLE child;
```

The output shows that the `child` table does not contain any foreign keys:

```sql
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| child | CREATE TABLE `child` (                                      |
|       |   `id` int(11) DEFAULT NULL,                                |
|       |   `pid` int(11) DEFAULT NULL                                |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```
````

Reviewer: @dveeden Don't use GitHub suggestion to commit here, since the last sentence was missed again.

> The above output of `SHOW CREATE TABLE` indicates that no foreign keys exist in the `child` table.

Author: @crazycs520 I think @Oreoxmt moved that to be above the output isn't it?

### Compatibility with MySQL

When you create a foreign key without specifying a name, the name generated by TiDB is different from that generated by MySQL. For example, the foreign key name generated by TiDB is `fk_1`, `fk_2`, and `fk_3`, while the foreign key name generated by MySQL is `table_name_ibfk_1`, `table_name_ibfk_2`, and `table_name_ibfk_3`.

> Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of constraint definitions are checked and enforced.

Reviewer: ```suggestion
Both MySQL and TiDB parse but ignore "inline `REFERENCES` specifications". Only `REFERENCES` specifications that are part of constraint definitions are checked and enforced. For example:

```sql
CREATE TABLE parent (
    id INT KEY
);

CREATE TABLE child (
    id INT,
    pid INT REFERENCES parent(id)
);

show create table child;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| child | CREATE TABLE `child` (                                      |
|       |   `id` int(11) DEFAULT NULL,                                |
|       |   `pid` int(11) DEFAULT NULL                                |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

In in uppercase, `SHOW CREATE TABLE` statement output indicates that no foreign keys exist in table `child`.
```
